src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,901 @@
+    DataModel::Nullable<TimeSynchronization::Structs::TrustedTimeSourceStruct::Type> tts;
+
+    if (!timeSource.IsNull())
+    {
+
+        TimeSynchronization::Structs::TrustedTimeSourceStruct::Type ts = { commandObj->GetAccessingFabricIndex(),
+                                                                           timeSource.Value().nodeID, timeSource.Value().endpoint };
+        tts.SetNonNull(ts);
+        // TODO schedule a utctime read from this time source
+    }
+    else
+    {
+        tts.SetNull();
+        sendMissingTrustedTimeSourceEvent(commandPath.mEndpointId);
+    }
+
+    TimeSynchronizationServer::Instance().SetTrustedTimeSource(tts);
+    commandObj->AddStatus(commandPath, Status::Success);
+    return true;
+}
+
+bool emberAfTimeSynchronizationClusterSetTimeZoneCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetTimeZone::DecodableType & commandData)
+{
+    const auto & timeZone = commandData.timeZone;
+    uint64_t localTime;
+
+    CHIP_ERROR err = TimeSynchronizationServer::Instance().SetTimeZone(timeZone);
+    if (err != CHIP_NO_ERROR)
+    {
+        if (err == CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            commandObj->AddStatus(commandPath, Status::ResourceExhausted);
+        }
+        else
+        {
+            commandObj->AddStatus(commandPath, Status::ConstraintError);
+        }
+        return true;
+    }
+    sendTimeZoneStatusEvent(commandPath.mEndpointId, 0);
+    sendTimeFailureEvent(commandPath.mEndpointId); // TODO remove
+    GetDelegate()->HandleTimeZoneChanged(TimeSynchronizationServer::Instance().GetTimeZone());
+    GetDelegate()->HandleDstoffsetlookup();
+
+    TimeSynchronization::TimeZoneDatabaseEnum tzDb;
+    TimeZoneDatabase::Get(commandPath.mEndpointId, &tzDb);
+    Commands::SetTimeZoneResponse::Type response;
+    if (GetDelegate()->HasFeature(commandPath.mEndpointId, TimeSynchronization::Feature::kTimeZone) &&
+        tzDb != TimeSynchronization::TimeZoneDatabaseEnum::kNone)
+    {
+        auto tzL = TimeSynchronizationServer::Instance().GetTimeZone();
+        auto tz  = tzL.begin();
+        if (tz != tzL.end() && tz->name.HasValue() && GetDelegate()->HandleDstoffsetavailable(tz->name.Value()))
+        {
+            GetDelegate()->HandleGetdstoffset();
+            response.DSTOffsetRequired = false;
+            sendDSTStatusEvent(commandPath.mEndpointId, true);
+        }
+        else
+        {
+            TimeSynchronizationServer::Instance().ClearDSTOffset();
+            sendDSTTableEmptyEvent(commandPath.mEndpointId);
+            sendDSTStatusEvent(commandPath.mEndpointId, false);
+            response.DSTOffsetRequired = true;
+        }
+    }
+    else
+    {
+        response.DSTOffsetRequired = true;
+    }
+    TimeSynchronizationServer::Instance().GetLocalTime(commandPath.mEndpointId, localTime);
```

Comment:
Ah. OK, yeah, this might just be weird spec wording. Since LocalTime does not support subscriptions, it effectively means that any reads to the attribute after this command should use this time zone, as well as any internal time references. 

Does GetLocalTime have any side effects that are important here? Or is it just a readout of the time? If this is really doing nothing and is just there to appease the spec wording, it's better to update the spec text.

Reply:
GetLocalTime doesn't have any side effects. It only picks the correct time zone and dst offset and calculate local time.